Fix for sending only failed data to kinesis on full or partial failed uploads

hackgnar opened the pull request:
@zwass could I get your opinion/assistance on this PR?  In short, there is an issue in the kinesis logger plugin where duplicate data gets sent to kinesis if there is an upload which partially fails.  

It may not seem too dire, but I have seen single endpoints send over 100k dup events over a 24h time period from just ~3k source events.  This typically happens in environments where kinesis is under heavy load and rate limit errors pop up.

With the current code path, when you receive errors like the following, osquery will resend all data (even the ones that successfully uploaded) until all data in the batch is successfully uploaded.

```
E1101 15:20:31.287334 4284416 aws_kinesis.cpp:111] Kinesis write for 5 of 140 records failed with error Rate exceeded for shard....REDACTED...
```

So in the above example, osquery tried to send 140 records.  135 of them were successfully inserted into kinesis but 5 failed. Osquery will resend all 140 records over and over again until all 140 are successfully upload.  When kinesis is under heavy load and rate limits are the cause of the failure, the next uploads will typically fail as well.  This can cause a large amount of dup data. 

Im not a huge fan of my PR, but I wanted to get something submitted sooner than later so we can agree on a proper fix.  It should also help you get an idea of the issue.

Do you think we should:

1. clean up this PR and go with it?
2. drop the flags and come up with sane defaults?
3. just delete the successful events from the log_data arg (possibly the best solution)
4. do something completely different?

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
Can one of the osquery admins reply with "ok to test" to kick off a build...

osqueryer commented:
:-1: The commit 9638d26c35b737d8a3fcfb1f64a75d94d94ca1a1 (Job results: [3596](https://jenkins.osquery.io/job/osqueryPullRequestBuild/3596/)) failed one or more tests (**OS X/Linux**).

osqueryer commented:
:no_good: The commit de9549297d7decdf02b240074a6094b20cba394a (Job results: [579](https://jenkins.osquery.io/job/osqueryPullRequestAudit/579/)) failed the code audit.

zwass commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -87,19 +95,36 @@ Status KinesisLogForwarder::send(std::vector<std::string>& log_data,
+      VLOG(1) << "Resending " << result.GetFailedRecordCount()
+              << " records to Kinesis";
+      retry_count = --retry_count;
+      send(resend, log_type);
```
I'm not thrilled about the (potentially deep) recursion here. Maybe it would feel better if it was a loop instead? That way we could make the retry count a local variable as well.

If we did go along with this approach, I think we need to be doing something with the return value of `send()` here.

theopolis replied:
Can you pull the heavy-lifting out of the `send` method? Then you can loop and call that new method here, without introducing any recursion. It would also remove the need for a "state" tracking variable (`retry_count`), making this a lot safer.

zwass commented:
Let me also note that I'm going to be mostly unavailable until the end of this weekend. Thank you for working on this!

hackgnar commented:
Thanks for the feedback @zwass .  Ill try to carve out some time in the next day or so and create/test/push a cleaner PR with your suggestions.

theopolis commented:
Can we make this feature a bit more seemless? Or at least overload/borrow from other flags? Is there a case where you'd need to change the introduced values?

hackgnar commented:
@theopolis I'll see what I can do.  I would also like to avoid more flags in the aws plugin.

hackgnar commented:
@zwass @theopolis ok, I changed the code around:
1. changed it to a while loop
2. removed the flags and replaced them with reasonable defaults with an exponential backoff delay
3. smoke tested it

Tomorrow, I'll 
1. clean up the code
2. test a bit more
3. make a unit test
4. send up the new PR 

You guys know the C++ better than me, so I figure you'll have some optimization changes for me to make.  Thx again guys =)

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -36,6 +36,14 @@ FLAG(uint64,
      10,
      "Seconds between flushing logs to Kinesis (default 10)");
+FLAG(uint64,
+     aws_kinesis_failed_upload_retry_count,
```
`aws_kinesis_retry`

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -36,6 +36,14 @@ FLAG(uint64,
+     aws_kinesis_failed_upload_retry_count,
+     10,
+     "How many times failed uploads will be reattempted");
+FLAG(uint64,
```
Nix this, just use the `aws_kinesis_period` with seconds as your precision.

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -87,19 +95,36 @@ Status KinesisLogForwarder::send(std::vector<std::string>& log_data,
   if (result.GetFailedRecordCount() != 0) {
+    std::vector<std::string> resend;
+    std::string error_msg = "";
+    int i = 0;
```
Unless you need this signed, prefer a `size_t`.

theopolis commented on osquery/logger/plugins/aws_kinesis.h:
```diff
@@ -47,6 +48,7 @@ class KinesisLogForwarder : public BufferedLogForwarder {
   std::shared_ptr<Aws::Kinesis::KinesisClient> client_{nullptr};
 
   FRIEND_TEST(KinesisTests, test_send);
+  int retry_count = FLAGS_aws_kinesis_failed_upload_retry_count;
```
This should be a `size_t`. It should also be initialized during `setUp` as the config may want to change the flag value. Right now it will only every be the default value.

theopolis replied:
Also, place it before the `FRIEND_TEST`. And if you don't mind, add doc-blocks for the methods in this class! :D

hackgnar commented:
@theopolis & @zwass I just sent up the code rewrite.  Im currently doing a thurough test on it but it will likely need some c++ cleanup.  Any suggestions are appreciated.  

Most of this rewrite nulls @theopolis suggestions from yesterday.

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
-      LOG(ERROR) << "Kinesis log too big, discarding!";
+  int retry_count = 100;
+  int retry_delay = 3000;
+  std::vector<std::string> batch = log_data;
```
This is potentially VERY expensive. `log_data` is passed to `::send` as a mutable reference, can we use that instead of making a copy?

hackgnar replied:
nm, this is simple... making changes now... PR update in T minus 60 min (need to run some smoke tests)

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
-  for (const std::string& log : log_data) {
-    if (log.size() > kKinesisMaxLogBytes) {
-      LOG(ERROR) << "Kinesis log too big, discarding!";
+  int retry_count = 100;
```
If these should not be signed, prefer `size_t`.

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
+    Aws::Kinesis::Model::PutRecordsResult result = outcome.GetResult();
+    VLOG(1) << "Successfully sent "
+            << result.GetRecords().size() - result.GetFailedRecordCount()
+            << " of " << result.GetRecords().size() << " logs to Kinesis.";
```
Avoid punctuation in log lines. :)

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
+    if (result.GetFailedRecordCount() != 0) {
+      std::vector<std::string> resend;
+      std::string error_msg = "";
+      int i = 0;
```
Do you intend for this to go below 0?

hackgnar replied:
Checking the api docs on this, but I don't think so.  This was just carry over code from the previous functionality.

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
+      // exit if all uploads fail right off the bat
+      // exit if we have tried too many times
+      // note, this will go back to the default logger batch retry code
+      if (retry_count < 1 ||
```
If this were an unsigned, it should be `if (retry_count > 0`.

hackgnar replied:
you mean `retry_count = 0` right?

theopolis commented on osquery/logger/plugins/aws_kinesis.cpp:
```diff
@@ -65,41 +65,69 @@ Status KinesisLoggerPlugin::logString(const std::string& s) {
+      // exit if we have tried too many times
+      // note, this will go back to the default logger batch retry code
+      if (retry_count < 1 ||
+          int(log_data.size()) == result.GetFailedRecordCount()) {
```
If you need to compare an int, to be save you'll need to make sure this wont overflow. Alternatively, checking `.GetFailedRecordCount` could be checked for `>=0`. In either case use a `static_cast`.

osqueryer commented:
:no_good: The commit 90abeebdac3265199f001aeba7e1747da001e60c (Job results: [645](https://jenkins.osquery.io/job/osqueryPullRequestAudit/645/)) failed the code audit.
